### PR TITLE
[halide-backend] Refactor codegen/triton.py into codegen/simd.py

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126417
* #126416
* __->__ #126415

This PR is primarily just moving stuff around.  It creates a new
common baseclass for TritonCodegen and the (upcoming) HalideCodegen.

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/simd.py [hunk 1]
```diff
@@ -0,0 +1,1916 @@
...
+        if isinstance(index, sympy.Integer):
+            expand_str = f"{copy_shape}.shape" if copy_shape else self.dense_size_str()
+            index_str = f"tl.full({expand_str}, {index_str}, tl.int32)"
+            return IndexingOptions(
+                index_str, set(), "None", expand_str, has_rindex, index
+            )
+
+        if need_dense and not have_dense:
+            expand_str = f"{copy_shape}.shape" if copy_shape else self.dense_size_str()
+            index_str = f"tl.broadcast_to({index_str}, {expand_str})"
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/codegen/simd.py:834]: I'll do the smaller refactor stuff like this in a followup PR.  This one is already large.

I tried to keep this one mainly as bulk code movement.  It is enough to get Halide codegen working down the stack.

**Review by shunting314 — APPROVED** [review]

**Pull request closed without merging**